crates/bevy_asset/src/processor/mod.rs

```diff
@@ -0,0 +1,1265 @@
+        let mut log = self.data.log.write().await;
+        let log = log.as_mut().unwrap();
+        log.begin_processing(path).await.unwrap();
+    }
+
+    /// Logs the end of an asset being successfully processed. See [`AssetProcessor::log_begin_processing`].
+    async fn log_end_processing(&self, path: &Path) {
+        let mut log = self.data.log.write().await;
+        let log = log.as_mut().unwrap();
+        log.end_processing(path).await.unwrap();
+    }
+
+    /// Starts the processor in a background thread.
+    pub fn start(_processor: Res<Self>) {
+        #[cfg(target_arch = "wasm32")]
+        error!("Cannot run AssetProcessor on WASM yet.");
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            let processor = _processor.clone();
+            std::thread::spawn(move || {
+                processor.process_assets();
+                futures_lite::future::block_on(processor.listen_for_source_change_events());
+            });
+        }
+    }
+
+    /// Processes all assets. This will:
+    /// * Scan the [`ProcessorTransactionLog`] and recover from any failures detected
+    /// * Scan the destination [`AssetProvider`] to build the current view of already processed assets.
+    /// * Scan the source [`AssetProvider`] and remove any processed "destination" assets that are invalid or no longer exist.
+    /// * For each asset in the `source` [`AssetProvider`], kick off a new "process job", which will process the asset
+    /// (if the latest version of the asset has not been processed).
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn process_assets(&self) {
+        let start_time = std::time::Instant::now();
+        debug!("Processing Assets");
+        IoTaskPool::get().scope(|scope| {
+            scope.spawn(async move {
+                self.initialize().await.unwrap();
+                let path = PathBuf::from("");
+                self.process_assets_internal(scope, path).await.unwrap();
+            });
+        });
+        // This must happen _after_ the scope resolves or it will happen "too early"
+        // Don't move this into the async scope above! process_assets is a blocking/sync function this is fine
+        futures_lite::future::block_on(self.finish_processing_assets());
+        let end_time = std::time::Instant::now();
+        debug!("Processing finished in {:?}", end_time - start_time);
+    }
+
+    /// Listens for changes to assets in the source [`AssetProvider`] and update state accordingly.
+    // PERF: parallelize change event processing
+    pub async fn listen_for_source_change_events(&self) {
+        debug!("Listening for changes to source assets");
+        loop {
+            let mut started_processing = false;
+
+            for event in self.data.source_event_receiver.try_iter() {
+                println!("{event:?}\n");
+                if !started_processing {
+                    // TODO: re-enable this after resolving state change signaling issue
```

Comment:
Ah right. This _was_ an issue before because the async_broadcast channel was configured to block when there was already a state queued in the channel. This meant that if there were no consumers of the `state` channel, it would block indefinitely when trying to set the state. I've since discovered this setting (and changed it), so this is fixed. I'll re-enable this line / remove the comment.